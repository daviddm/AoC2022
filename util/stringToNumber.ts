export const stringToNumber = (list: string[]) => {
  return list.map((str) => parseInt(str, 10));
};
