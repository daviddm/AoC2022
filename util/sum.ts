export const sum = (list: number[]) =>
  list.reduce((prev, curr) => prev + curr, 0);
