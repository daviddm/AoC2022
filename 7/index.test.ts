import { readFile } from "@util/readFile";
import { stringToNumber } from "@util/stringToNumber";
import { sum } from "@util/sum";
import { parseOutput, findFolders, calcSizes } from "./index";

describe("7", () => {
  let testList: string[];
  let list: string[];

  beforeEach(async () => {
    testList = await readFile("./7/test.txt");
    list = await readFile("./7/input.txt");
  });

  describe("test", () => {
    it("create file", async () => {
      expect.assertions(1);
      const result = await parseOutput(["$ ls", "14848514 b.txt"]);
      expect(result).toMatchSnapshot();
    });
    it("create dir", async () => {
      expect.assertions(1);
      const result = await parseOutput(["$ ls", "dir a"]);
      expect(result).toMatchSnapshot();
    });
    it("create dir with file in", async () => {
      expect.assertions(1);
      const result = await parseOutput([
        "$ ls",
        "dir a",
        "$ cd a",
        "$ ls",
        "14848515 b.txt",
      ]);
      expect(result).toMatchSnapshot();
    });
    it("find folders", async () => {
      expect.assertions(1);
      const result = await findFolders(
        parseOutput(["$ ls", "dir a", "$ cd a", "$ ls", "1337 b.txt"])
      );
      expect(result).toMatchSnapshot();
    });
    it("calcSizes", async () => {
      expect.assertions(1);
      const root = await parseOutput([
        "$ ls",
        "dir a",
        "$ cd a",
        "$ ls",
        "1337 b.txt",
      ]);

      const size = calcSizes(root);
      expect(root.size).toBe(1337);
    });
  });

  describe("first", () => {
    it("structure", async () => {
      expect.assertions(1);
      const result = await parseOutput(testList);
      expect(result).toMatchSnapshot();
    });
    it("folders", async () => {
      expect.assertions(1);
      const result = await findFolders(parseOutput(testList));
      expect(result.length).toEqual(4);
    });
    it("size of folder e", async () => {
      expect.assertions(1);
      const root = parseOutput(testList);
      calcSizes(root);
      const dirs = findFolders(root).filter((a) => a.name == "e");
      expect(dirs[0].size).toEqual(584);
    });
    it("size of folder a", async () => {
      expect.assertions(1);
      const root = parseOutput(testList);
      calcSizes(root);
      const dirs = findFolders(root).filter((a) => a.name == "a");
      expect(dirs[0].size).toEqual(94853);
    });
    it("size of root", async () => {
      expect.assertions(1);
      const root = parseOutput(testList);
      calcSizes(root);
      const dirs = findFolders(root).filter((a) => a.name == "/");
      expect(dirs[0].size).toEqual(48381165);
    });

    it("test", async () => {
      expect.assertions(1);
      const root = parseOutput(testList);
      calcSizes(root);
      const dirSum = findFolders(root)
        .filter((a) => a.size <= 100000)
        .map((a) => a.size)
        .reduce(sum, 0);
      expect(dirSum).toEqual(95437);
    });

    it("answer", async () => {
      expect.assertions(1);
      const root = parseOutput(list);
      calcSizes(root);
      const dirSum = findFolders(root)
        .filter((a) => a.size <= 100000)
        .map((a) => a.size)
        .reduce(sum, 0);
      expect(dirSum).toEqual(1581595);
    });
  });

  describe("second", () => {
    it("test", async () => {
      expect.assertions(1);
      const root = parseOutput(testList);
      calcSizes(root);
      const min = Math.abs(70000000 - 30000000 - root.size);
      const dir = findFolders(root)
        .sort((a, b) => a.size - b.size)
        .find((a) => a.size > min);
      expect(dir?.size).toEqual(24933642);
    });
    it("answer", async () => {
      expect.assertions(1);
      const root = parseOutput(list);
      calcSizes(root);
      const min = Math.abs(70000000 - 30000000 - root.size);
      const dir = findFolders(root)
        .sort((a, b) => a.size - b.size)
        .find((a) => a.size > min);

      expect(dir?.size).toEqual(1544176);
    });
  });
});
