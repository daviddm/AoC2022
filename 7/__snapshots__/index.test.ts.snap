// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`7 first structure 1`] = `
{
  "directories": [
    {
      "directories": [
        {
          "directories": [],
          "files": [
            {
              "name": "i",
              "size": 584,
            },
          ],
          "name": "e",
          "parent": [Circular],
          "size": 0,
        },
      ],
      "files": [
        {
          "name": "f",
          "size": 29116,
        },
        {
          "name": "g",
          "size": 2557,
        },
        {
          "name": "h.lst",
          "size": 62596,
        },
      ],
      "name": "a",
      "parent": [Circular],
      "size": 0,
    },
    {
      "directories": [],
      "files": [
        {
          "name": "j",
          "size": 4060174,
        },
        {
          "name": "d.log",
          "size": 8033020,
        },
        {
          "name": "d.ext",
          "size": 5626152,
        },
        {
          "name": "k",
          "size": 7214296,
        },
      ],
      "name": "d",
      "parent": [Circular],
      "size": 0,
    },
  ],
  "files": [
    {
      "name": "b.txt",
      "size": 14848514,
    },
    {
      "name": "c.dat",
      "size": 8504156,
    },
  ],
  "name": "/",
  "parent": [Circular],
  "size": 0,
}
`;

exports[`7 test create dir 1`] = `
{
  "directories": [
    {
      "directories": [],
      "files": [],
      "name": "a",
      "parent": [Circular],
      "size": 0,
    },
  ],
  "files": [],
  "name": "/",
  "parent": [Circular],
  "size": 0,
}
`;

exports[`7 test create dir with file in 1`] = `
{
  "directories": [
    {
      "directories": [],
      "files": [
        {
          "name": "b.txt",
          "size": 14848515,
        },
      ],
      "name": "a",
      "parent": [Circular],
      "size": 0,
    },
  ],
  "files": [],
  "name": "/",
  "parent": [Circular],
  "size": 0,
}
`;

exports[`7 test create file 1`] = `
{
  "directories": [],
  "files": [
    {
      "name": "b.txt",
      "size": 14848514,
    },
  ],
  "name": "/",
  "parent": [Circular],
  "size": 0,
}
`;

exports[`7 test find folders 1`] = `
[
  {
    "directories": [
      {
        "directories": [],
        "files": [
          {
            "name": "b.txt",
            "size": 1337,
          },
        ],
        "name": "a",
        "parent": [Circular],
        "size": 0,
      },
    ],
    "files": [],
    "name": "/",
    "parent": [Circular],
    "size": 0,
  },
  {
    "directories": [],
    "files": [
      {
        "name": "b.txt",
        "size": 1337,
      },
    ],
    "name": "a",
    "parent": {
      "directories": [
        [Circular],
      ],
      "files": [],
      "name": "/",
      "parent": [Circular],
      "size": 0,
    },
    "size": 0,
  },
]
`;
